explorer/update.go

```diff
@@ -0,0 +1,189 @@
+	}
+
+	for _, elem := range cau.SpentSiacoins {
+		if err := e.db.RemoveElement(elem.ID); err != nil {
+			return err
+		}
+		facts.SpentSiacoinsCount++
+	}
+	for _, elem := range cau.SpentSiafunds {
+		if err := e.db.RemoveElement(elem.ID); err != nil {
+			return err
+		}
+		facts.SpentSiafundsCount++
+	}
+	for _, elem := range cau.ResolvedFileContracts {
+		if err := e.db.RemoveElement(elem.ID); err != nil {
+			return err
+		}
+		facts.ResolvedFileContractsCount++
+		facts.ActiveContractCount--
+		payout := elem.FileContract.MissedHostValue.Add(elem.FileContract.TotalCollateral)
+		facts.ActiveContractCost = facts.ActiveContractCost.Sub(payout)
+		facts.ActiveContractSize -= elem.FileContract.Filesize
+	}
+
+	for _, elem := range cau.NewSiacoinElements {
+		if err := e.db.AddSiacoinElement(elem); err != nil {
+			return err
+		}
+		facts.NewSiacoinsCount++
+	}
+	for _, elem := range cau.NewSiafundElements {
+		if err := e.db.AddSiafundElement(elem); err != nil {
+			return err
+		}
+		facts.NewSiafundsCount++
+	}
+	for _, elem := range cau.RevisedFileContracts {
+		if err := e.db.AddFileContractElement(elem); err != nil {
+			return err
+		}
+		facts.RevisedFileContractsCount++
+		facts.TotalContractSize += elem.FileContract.Filesize
+		facts.TotalRevisionVolume += elem.FileContract.Filesize
+	}
+	for _, elem := range cau.NewFileContracts {
+		if err := e.db.AddFileContractElement(elem); err != nil {
+			return err
+		}
+		facts.NewFileContractsCount++
+		payout := elem.FileContract.MissedHostValue.Add(elem.FileContract.TotalCollateral)
+		facts.ActiveContractCost = facts.ActiveContractCost.Add(payout)
+		facts.ActiveContractSize += elem.FileContract.Filesize
+		facts.TotalContractCost = facts.TotalContractCost.Add(payout)
+		facts.TotalContractSize += elem.FileContract.Filesize
+	}
+
+	if err := e.db.AddBlockFacts(facts.Index.Height, facts); err != nil {
+		return err
+	}
+	if err := e.db.SetBlockHeight(facts.Index.Height); err != nil {
+		return err
+	}
+
+	e.vc = cau.Context
+	return nil
+}
+
+// ProcessChainRevertUpdate implements chain.Subscriber.
+func (e *Explorer) ProcessChainRevertUpdate(cru *chain.RevertUpdate) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	for _, elem := range cru.SpentSiacoins {
+		if err := e.db.AddSiacoinElement(elem); err != nil {
+			return err
+		}
+	}
+	for _, elem := range cru.SpentSiafunds {
+		if err := e.db.AddSiafundElement(elem); err != nil {
+			return err
+		}
+	}
+	for _, elem := range cru.ResolvedFileContracts {
+		if err := e.db.AddFileContractElement(elem); err != nil {
+			return err
+		}
+	}
+
+	for _, elem := range cru.NewSiacoinElements {
+		if err := e.db.RemoveElement(elem.ID); err != nil {
+			return err
+		}
+	}
+	for _, elem := range cru.NewSiafundElements {
+		if err := e.db.RemoveElement(elem.ID); err != nil {
+			return err
+		}
+	}
+	for _, elem := range cru.RevisedFileContracts {
+		if err := e.db.RemoveElement(elem.ID); err != nil {
```

Comment:
the contract still exists even if a revision is reverted. It should be reset to its prior state (`elem.Parent`)